### PR TITLE
Fix enumerate parallel via DiCE

### LINKED ISSUE #846: Errors when combining iarange + discrete enumeration
This issue tracks progress of combining enumeration #742 and nested `iarange` #780. Currently `TraceEnum_ELBO` gives incorrect results when enumerating any `iarange`, even in SS-VAE (as discovered by @rohitsingh0812).

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #846 :tada:
Addresses #847

- Simplifies `TraceEnum_ELBO` to use the [DiCE](https://arxiv.org/abs/1802.05098) gradient estimator.
- Completely fixes parallel + sequential + MCMC enumeration logic (in all combinations).

Joint work with @eb8680 who wrote the [first version](https://github.com/uber/pyro/blob/dice-elbo/pyro/infer/dice_elbo.py) of `Dice_ELBO` from which much of this is copied.

### CODE REVIEW

**Comment by eb8680** [comment]
Nice!  FYI my less-tested version with extra Rao-Blackwellisation is [here](https://github.com/uber/pyro/blob/dice-elbo/pyro/infer/dice_elbo.py).

File: pyro/infer/util.py [hunk 1]
```diff
@@ -124,68 +125,71 @@ def __repr__(self):
...
-    Data structure to compute cumulative costs along paths in a tree.
-    Typically keys are ``cond_indep_stack``s.
+    An implementation of the DiCE operator compatible with Pyro features.
+
+    This implementation correctly handles:
+    - scaled log-probability due to subsampling
+    - independence in different contexts due to iarange
+    - weights due to parallel and sequential enumeration
+
+    This assumes restricted dependency structure on the model and guide:
```
Comment (martinjankowiak) [inline pyro/infer/util.py]: "This assumes restricted dependency..."

are we going to add checks for this?

File: pyro/infer/util.py [hunk 2]
```diff
@@ -124,68 +125,71 @@ def __repr__(self):
...
-    def copy(self):
-        result = TreeSum()
-        result._terms = self._terms.copy()
-        result._upstream = self._upstream.copy()
-        result._frozen = self._frozen
-        return result
+        for site in guide_trace.nodes.values():
+            if site["type"] != "sample":
+                continue
+            log_prob = site['score_parts'].score_function  # not scaled by subsampling
```
Comment (martinjankowiak) [inline pyro/infer/util.py]: can we add a test with enumeration with rejection_gamma?

File: pyro/infer/util.py (hunk 1 above)
Comment (fritzo) [reply pyro/infer/util.py]: We can add some partial checks before release, but some cases are impossible to detect. E.g. in
```py
x_axis = iarange('x')
y_axis = iarange('y')
with x_axis:
    x = pyro.sample('x', ...)
with y_axis:
    y = pyro.sample('y', ...)
with x_axis, y_axis:
    xy = pyro.sample("xy", ...)
```
it is impossible to determine whether `y` depends on `x`. (we could examine the autograd graph, but the user could get around this via `x.detach()`).

**Comment by martinjankowiak** [comment]
another question i have is the following;

let's say i only wanted it to be once differentiable. would it be simple to support that case and if so would it lead to a a performance gain? seems like the answer should be yes on both points?

**Comment by fritzo** [comment]
> let's say i only wanted it to be once differentiable

I think you could wrap it in a `torch.autograd.Function` with a `@once_differentiable` `.backward()`.

Actually I think we could be more efficient by wrapping the entire `(log_prob - log_prob.detach()).exp()` op in a `torch.autograd.Function` that avoids computing `log_prob`, but that seems relatively low priority.

**Comment by fritzo** [comment]
> i'm unclear if rsvi will work correctly

I believe RSVI will work correctly, but will fail to take advantage of sticking-the-landing.

**Comment by martinjankowiak** [comment]
wouldn't it be sufficient to add a detach in one or two places if ```only_once_differentiable=True```?

**Comment by fritzo** [comment]
> let's say i only wanted it to be once differentiable
> wouldn't it be sufficient to add a detach

Actually the way we're using this in `loss_and_grads` should already be nearly as cheap as a once-differentiable version, since we're not passing `create_graph=True` to `.backward()`. In that sense, we're only using it in a once-differentiable way.
